exports-last

k15a opened the pull request:
Hey,

I am currently working on #620 and would like some feedback.

Currently my implementation only checks for ES Modules export declaration but I would also like to support CommonJS modules as well. Does anybody has an idea how to check if an ExpressionStatement is an CommonJS export?

benmosher commented:
The ES module part looks good to me. 👍 

I'd leave out CommonJS for now and add it later behind a flag (unless it's a use case you're looking forward to linting).

k15a commented:
I've just added documentation for this rule. 📖

k15a commented:
I've resolved the merge conflicts but now travis ci is failing and I couldn't figure out why.

k15a commented:
I've removed the CommonJS part. Let me know if I can do something else to get this merged.

benmosher commented:
LGTM, but would like to have one of @jfmengels or @ljharb take a quick look and sign off as well before merging 👍🏻

jfmengels commented on src/rules/exports-last.js:
```diff
@@ -0,0 +1,35 @@
+function isExportStatement({ type }) {
+  // ES Module export statements
+  if (type === 'ExportDefaultDeclaration' || type === 'ExportNamedDeclaration') {
```
Simpler:
```js
function isExportStatement({ type }) {
  return type === 'ExportDefaultDeclaration' || type === 'ExportNamedDeclaration';
}
```

And you should probably also test for ExportAllDeclaration (and add tests for that case)

jfmengels commented on src/rules/exports-last.js:
```diff
@@ -0,0 +1,35 @@
+  create(context) {
+    return {
+      Program({ body }) {
+        const lastNonExportStatement = body.reduce((acc, node, index) => {
```
rename to `lastNonExportStatementIndex`

jfmengels commented on src/rules/exports-last.js:
```diff
@@ -0,0 +1,35 @@
+
+        body.forEach((node, index) => {
+          if (isExportStatement(node) && index < lastNonExportStatement) {
+
```
remove this blank line (sorry about the nitpicking :sweat_smile:)

jfmengels commented on src/rules/exports-last.js:
```diff
@@ -0,0 +1,35 @@
+        }, 0)
+
+        body.forEach((node, index) => {
+          if (isExportStatement(node) && index < lastNonExportStatement) {
```
It might be a bit faster to do the `index < lastNonExportStatement` before the `isExportStatement`.

Or even better (?): only run the `forEach` on the statements after `lastNonExportStatement`.

```js
// maybe add +/-1 to lastNonExportStatement to make it work right
body.slice(lastNonExportStatement).forEach(node => {
  if (isExportStatement(node)) {
    // report
  }
})
```

jfmengels commented on docs/rules/exports-last.md:
```diff
@@ -0,0 +1,41 @@
+
+## When Not To Use It
+
+If you don't mind exports being sprinkled throughout a file, you may not want to enable this rule.
```
It would be nice to indicate that this is only for the export statements using the `export` keyword, and not for CommonJS `module.exports/exports`

k15a commented:
Hey @jfmengels, I've added some more tests unfortunately one test is breaking but I think that's not an issue with my rule. I've also refactored the code. Let me know if I can do something else.

k15a commented on tests/src/rules/exports-last.js:
```diff
@@ -0,0 +1,117 @@
+  valid: [
+    // Empty file
+    test({
+      code: '',
```
This line is breaking the tests.

jfmengels replied:
Ok, try replacing it with a comment
```js
test({
        code: '// comment',
})
```
and if that doesn't work out, just remove the test case.

jfmengels commented on docs/rules/exports-last.md:
```diff
@@ -0,0 +1,48 @@
+# exports-last
+
+This rule reports all export declaration which come before any non-export statements.
```
`This rule enforces that all exports are declared at the bottom of the file. This rule will report any export declarations that comes before any non-export statements.`

jfmengels commented on tests/src/rules/exports-last.js:
```diff
@@ -0,0 +1,117 @@
+        export const even = 'count'
+        export const how = 'many'
+      `,
+      errors,
```
Shouldn't there be 2 errors here? One for the `export default` and one for ` export const so = 'many'`?

jfmengels commented on tests/src/rules/exports-last.js:
```diff
@@ -0,0 +1,117 @@
+
+const ruleTester = new RuleTester()
+
+const errors = ['Export statements should appear at the end of the file']
```
```js
const error = {
  ruleId: 'exports-last',
  message: 'Export statements should appear at the end of the file',
};
```
You can make this into an array by default but I think you'll need to make an array of multiple errors for some tests.

Would be nice if you could add:
- either the location of the reported node
- the type of the Node : `type: 'ImportDefaultSpecifier`. This might be easier.

At the moment, you are not reporting the export statements, but non-export statements. Any would be fine (though I' prefer reporting the export statements), but the message mentions the export statement, so let's report those. That's why I'd like to add this additional verification.

It could look something like:
```js
const error = type => {
  ruleId: 'exports-last',
  message: 'Export statements should appear at the end of the file',
  type
};

// ------

     test({
        code: `
          export * from './foo'
          const bar = true
        `,
        errors: [error('ExportAllDeclaration')],
      }),
```

jfmengels commented on src/rules/exports-last.js:
```diff
@@ -0,0 +1,27 @@
+  },
+}
+
+export default rule
```
Would you mind making the file structure (exports and stuff) look a bit more like the others, for consistency? Here's an example: https://github.com/benmosher/eslint-plugin-import/blob/master/src/rules/no-dynamic-require.js

jfmengels commented on CHANGELOG.md:
```diff
@@ -26,6 +26,9 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 - Fixed documentation for the default values for the [`order`] rule ([#601])
 
+### Added
+- [`exports-last`] lints that export statements are at the end of the file ([#620] + [#632])
```
You'll need to rebase and move this back into the `Unreleased` section. Things have changed ever since you mad your PR :sweat_smile: 

If you don't know how to rebase, don't mind it, I'll do it later for you.

jfmengels commented on docs/rules/exports-last.md:
````diff
@@ -0,0 +1,48 @@
+## This will not be reported
+
+```JS
+export const bool = true
````
Add a simple non-exported variable declaration up top, to show that that is allowed

k15a commented:
Hey @jfmengels,

I think I know solved all the requested changes. Let me know if I can do something else to get this merged.

ljharb commented:
@k15a please rebase latest master, so as to remove all merge commits

k15a commented:
I tried my best. Did this removed the merge commits?

ljharb commented on CHANGELOG.md:
```diff
@@ -6,8 +6,8 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 ### Added
 - [`no-anonymous-default-export`] rule: report anonymous default exports ([#712], thanks [@duncanbeevers]).
-- Add new value to [`order`]'s `newlines-between` option to allow newlines inside import groups ([#627], [#628], thanks [@giodamelio])
-- Add `count` option to the [`newline-after-import`] rule to allow configuration of number of newlines expected ([#742], thanks [@ntdb])
```
we lost one of the changelog entries - let's fix that :-)

k15a commented:
Fixed that as well. Hope everything is ok now.

alexkrolick commented:
I'd like to use this rule - anyone want to take a look at it again?

sevor005 commented:
Good afternoon! The question is: can I cancel the rule (put it in ignor) for classes that are exported at the beginning of the file export default class SomeClass, followed by defaultProps and PropTypes?
